06_IPC/03_Message_Passing.md

```diff
@@ -0,0 +1,176 @@
+
+- What happens if there unread messages when destroying an endpoint? How do you handle them?
+- Who is allowed to remove an endpoint?
+
+## Sending A Message
+
+Now that we know where to send the data, let's look at the process for that.
+
+When a process has a message it wants to send to the endpoint, it writes it into a buffer. We then tell the IPC manager that we want to send this buffer to this endpoint. This hints at what our function prototype might look like:
+
+```c
+void ipc_send(void* buffer, size_t length, const char* ep_name);
+```
+
+The `ep_name` argument is the name of the endpoint we want to send to in this case. Which leads nicely into the first thing we'll need to do: find an endpoint in the list with the matching name. This is a pretty standard algorithm, we just loop over each element comparing the endpoint's name with the one we're looking for.
+
+```c
+ipc_endpoint* target = first_endpoint;
+//search the list for the endpoint we want
+acquire(&endpoints_lock);
+while (target != NULL) {
+    if (strcmp(target->name, ep_name) == 0)
+        break;
+    target = target->next;
+}
+
+release(&endpoints_lock);
+if (target == NULL)
+    return;
+```
+
+You may want to return an error here if the endpoint couldn't be found, however in our case we're simply discarding the message. 
+
+Now we'll need to allocate a buffer to store a copy of the message in, and copy the original message into this buffer.
+
+```c
+void* msg_copy = malloc(length);
+memcpy(msg_copy, buffer, length);
+```
+
+Why do we make a copy of the original message? Well if we dont, the sending process has to keep the original message around until the receiver has processed it. We don't have a way for the receiver to communicate that it's finished reading the message. By making a copy, we can return from `ipc_send()` as soon as the message is sent, regardless of when the message is read. Now the sending process is free to do what it wants with the memory holding the original message as soon as `ipc_send()` has completed.
+
+If you're performing this IPC as part of a system call from userspace, the memory containing the original message is unlikely to be mapped in the receiver's address space anyway, so we have to copy it into the kernel's address space, which is mapped in both processes.
+
+All that's left is to tell the receiver it has a message is available by placing the buffer address on the endpoint. Again, notice the use of the lock to prevent race conditions while we mess with the internals of the endpoint.
+
+```c
+acquire(&target->lock);
+target->msg_buffer = msg_copy;
+target->msg_length = length;
+release(&target->lock);
+```
+
+After the lock on the endpoint is released, the sent has been sent! Now it's up to the receiving thread to check the endpoint and set the buffer to NULL again.
```

Comment:
the sent has been sent? 

pun intended or typo?

Reply:
type :)